src/_png.cpp

```diff
@@ -287,10 +295,23 @@ static PyObject *_read_png(PyObject *filein, bool float_result)
         py_file = filein;
     }
 
-    if ((fp = mpl_PyFile_Dup(py_file, (char *)"rb", &offset))) {
+    #if PY3K
+    if (close_file) {
+    #else
+    if (close_file || PyFile_Check(py_file)) {
+    #endif
+        fp = mpl_PyFile_Dup(py_file, (char *)"rb", &offset);
+    }
```

Comment:
hopefully, this won't make merges down to master more difficult later, since master already has a version of this change (just with the nbagg optimizations).